### PR TITLE
fix: resolve strict concurrency errors

### PR DESCRIPTION
**Pull request by phantumcode**
## Description
Resolving errors in various categories when strict concurrency and Swift 6 experimental settings are enabled
### Change details:

**Amplify Core**
* Updated usage and references to `UIDevice` in `DeviceInfo` from an main-actor isolated context
* Updated code in `DevMenu` resolve issues with referencing main-actor isolated context from a synchronus nonisolated context

**AmplifyPlugins/Core**
* Updated `AWSServiceConfiguration` to calculate the user agent value in an main-actor isolated context as a result of changes to Amplify Core

**AmplifyPlugins Auth**
* Changes from AmplifyCore required cascading changes to AmplifyPlugins Auth due to `ASFDeviceInfo` referencing `DeviceInfo`
* Updated calling hierarchy to reference `ASFDeviceInfo` in an async isolated context

**AmplifyPlugins InternalAWSPinpoint**
* Changes from AmplifyCore required cascading changes to `EndpointInformation`
* Updated `EndpointInformation` to be a struct and introduced `EndpointInformationProvider` interface and implementation so that calling hierarchy can execute and get `EndpointInformation` in an async isolated context

**AmplifyPlugins Logging**
* The `CloudWatchLoggingStreamNameFormatter` needs to calculate the device identifier to calculate the log stream name in an async isolated context.  Updated and refactor logging logic to calculate the log stream name at the time of usage so that it can be called in an async isolated context while limiting cascading changes.

**AmplifyPlugins Storage**
* The Storage plugin references `AWSServiceConfiguration` in AmplifyPluginsCore which requires main-actor isolated context execution.  Updating the `StorageServiceProxy` to provide `userAgent` computed property in an async context.  Implementation and references to `StorageServiceProxy` to get the `userAgent` property is also updated to execute in an async context.

### Testing details:
* verified all unit tests passed when running with Xcode 15.0.1 and Xcode 15.4
* verified all integration tests passed when running with Xcode 15.0.1 and Xcode 15.4

## General Checklist

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [x] All integration tests pass
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [x] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [x] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [x] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Context/PinpointContext.swift [hunk 1]
```diff
@@ -110,7 +110,7 @@ class PinpointContext {
     private let storage: PinpointContextStorage
 
     init(with configuration: PinpointContextConfiguration,
-         endpointInformation: EndpointInformation = .current,
+         endpointInformationProvider: EndpointInformationProvider = DefaultEndpointInformationProvider(),
```
Comment (thisisabhash) [inline AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Context/PinpointContext.swift:113]: is `EndpointInformation.current` the same as `DefaultEndpointInformationProvider` ?

File: AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift [hunk 2]
```diff
@@ -8,26 +8,24 @@
 import Amplify
 import Foundation
 
-protocol EndpointInformation {
+struct EndpointInformation {
```
Comment (thisisabhash) [inline AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift:11]: why this change? or is it because of rebase?

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/AWSCloudWatchLoggingSessionController.swift [hunk 3]
```diff
@@ -115,9 +114,9 @@ final class AWSCloudWatchLoggingSessionController {
         }
 
         guard let cloudWatchClient = client else { return nil }
-        return try CloudWatchLoggingConsumer(client: cloudWatchClient,
-                                             logGroupName: self.logGroupName,
-                                             userIdentifier: self.userIdentifier)
+        return CloudWatchLoggingConsumer(client: cloudWatchClient,
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/AWSCloudWatchLoggingSessionController.swift:117]: nit: indentation

File: AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Context/PinpointContext.swift (hunk 1 above)
Comment (phantumcode) [reply AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Context/PinpointContext.swift:113]: no, the provider will provide the current endpoint when it's needed, deferring the execution to get the current endpoint information to a latter point when we're already in an async context.  This avoids the need to make the PinpointContext constructor async

File: AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift (hunk 2 above)
Comment (phantumcode) [reply AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift:11]: the original intent was to have `DeviceInfo` implement the `EndpointProtocol`, this simplifies it by having the provider return the struct that is constructed from `DeviceInfo`

File: Package.swift [hunk 4]
```diff
@@ -89,6 +89,7 @@ let amplifyTargets: [Target] = [
         dependencies: [],
         path: "AmplifyAsyncTesting/Sources/AsyncTesting",
         linkerSettings: [.linkedFramework("XCTest")]
+
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply Package.swift]: file reverted

File: Amplify/DevMenu/Trigger/LongPressGestureRecognizer.swift [hunk 5]
```diff
@@ -47,13 +47,5 @@ class LongPressGestureRecognizer: NSObject, TriggerRecognizer, UIGestureRecogniz
         uiWindow?.addGestureRecognizer(recognizer)
         recognizer.delegate = self
     }
-
-    deinit {
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply Amplify/DevMenu/Trigger/LongPressGestureRecognizer.swift:51]: yes, it's also not possible to remove the gesture from the window without an async context.  I'd like to propose marking DevMenu for deprecation outside of this PR since it is only documented for V1

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/RefreshAuthorizationSession/UserPool/RefreshUserPoolTokens.swift [hunk 6]
```diff
@@ -41,7 +41,7 @@ struct RefreshUserPoolTokens: Action {
                 for: existingSignedIndata.username,
                 credentialStoreClient: authEnv.credentialsClient)
 
-            let input = InitiateAuthInput.refreshAuthInput(
+            let input = await InitiateAuthInput.refreshAuthInput(
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/RefreshAuthorizationSession/UserPool/RefreshUserPoolTokens.swift:44]: it's due to the cascading async context requirement as a result of changes to `ASFDeviceInfo` and `DeviceInfo`

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Consumer/CloudWatchLoggingConsumer.swift [hunk 7]
```diff
@@ -75,6 +76,10 @@ extension CloudWatchLoggingConsumer: LogBatchConsumer {
             ensureLogStreamExistsComplete = true
         }
 
+        if logStreamName == nil {
+            self.logStreamName = await formatter.formattedStreamName()
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Consumer/CloudWatchLoggingConsumer.swift:80]: it needs to be created once and re-used again with the same logstream name based on device identifier and datetime.

File: AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginHostApp/AWSCloudWatchLoggingPluginIntegrationTests/AWSCloudWatchLoggingPluginIntegrationTests.swift [hunk 8]
```diff
@@ -61,7 +61,14 @@ class AWSCloudWatchLoggingPluginIntergrationTests: XCTestCase {
         await Amplify.reset()
         let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first?.path ?? NSTemporaryDirectory()
         let directory = documents.appendingPathComponent("amplify")
-                                 .appendingPathComponent("logging")
+            .appendingPathComponent("logging")
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginHostApp/AWSCloudWatchLoggingPluginIntegrationTests/AWSCloudWatchLoggingPluginIntegrationTests.swift]: Fixed indentation

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/AWSCloudWatchLoggingSessionController.swift (hunk 3 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/AWSCloudWatchLoggingSessionController.swift]: Fixed indentation

**Review by thisisabhash — DISMISSED** [review]

File: AmplifyPlugins/Internal/Tests/InternalAWSPinpointUnitTests/ActivityTrackerTests.swift [hunk 9]
```diff
@@ -79,7 +82,7 @@ class ActivityTrackerTests: XCTestCase {
         
         NotificationCenter.default.post(Notification(name: Self.applicationDidMoveToBackgroundNotification))
         NotificationCenter.default.post(Notification(name: Self.applicationWillMoveToForegoundNotification))
-        NotificationCenter.default.post(Notification(name: Self.applicationWillTerminateNotification))
+        await NotificationCenter.default.post(Notification(name: Self.applicationWillTerminateNotification))
```
Comment (ruisebas) [inline AmplifyPlugins/Internal/Tests/InternalAWSPinpointUnitTests/ActivityTrackerTests.swift:85]: 🤔 is the `await` really only needed for this particular notification?

File: AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift [hunk 10]
```diff
@@ -8,26 +8,24 @@
...
 }
 
-extension EndpointInformation where Self == DeviceInfo {
-    static var current: EndpointInformation {
-        DeviceInfo.current
+struct DefaultEndpointInformationProvider: EndpointInformationProvider {
+    func endpointInfo() async -> EndpointInformation {
+        let deviceInfo = await DeviceInfo.current
+        let model = await DeviceInfo.current.model
+        let platform = await DeviceInfo.current.operatingSystem
```
Comment (ruisebas) [inline AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift:27]:
```suggestion
        let model = await deviceInfo.model
        let platform = await deviceInfo.operatingSystem
```

**Review by ruisebas — APPROVED** [review]

File: AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift (hunk 10 above)
Comment (phantumcode) [reply AmplifyPlugins/Internal/Sources/InternalAWSPinpoint/Endpoint/EndpointInformation.swift]: Updated per suggestion.

File: AmplifyPlugins/Internal/Tests/InternalAWSPinpointUnitTests/ActivityTrackerTests.swift (hunk 9 above)
Comment (phantumcode) [reply AmplifyPlugins/Internal/Tests/InternalAWSPinpointUnitTests/ActivityTrackerTests.swift:85]: Yes, it's only one notification.  It's possible that this could change when Swift 6 is officially released.

**Review by lawmicha — COMMENTED** [review]
A reminder to merge this in as a `fix` or `feat` since `chore` commits do not show up in the changelog, and they usually non-production related code changes https://devops.stackexchange.com/a/19067
